### PR TITLE
Refactoring, typescript, images

### PR DESCRIPTION
**Pull request by mrquincle**
Refactoring, typescript, images

+ Split out into multiple files (cache, mapper, syncer, handler)
+ Transition into typescript
+ Update images following https://apps-sdk-v3.developer.athom.com/tutorial-App%20Store.html

Addresses most/all remarks by reviewer @AlexDM0

### CODE REVIEW

File: app.ts [hunk 1]
```diff
@@ -0,0 +1,377 @@
...
+		try {
+			await this.setupConnections(email, password);
+		}
+		catch(e) {
+			console.log('There was a problem making the connections:', e);
+			return;
+		};
+
+		console.log('Obtain all data from the cloud');
+		await this.syncer.getSpheres();
```
Comment (AlexDM0) [inline app.ts:92]: This could be the getAll as mentioned in the syncer class

File: app.ts [hunk 2]
```diff
@@ -0,0 +1,377 @@
...
+			if (data.subType === 'enterLocation') {
+				this.runLocationTrigger(data, true);
+			}
+			if (data.subType === 'exitLocation') {
+				this.runLocationTrigger(data, false);
+			}
+		}
+		// deletion and addition of Crownstones
+		if (data.type === 'dataChange') {
+			if (data.subType === 'stones') {
```
Comment (AlexDM0) [inline app.ts:217]: Don't we handle the create and trigger a getStones?

File: drivers/crownstone/driver.js [hunk 3]
```diff
@@ -14,7 +14,7 @@ class CrownstoneDriver extends Homey.Driver {
 	 * This method is called when the Driver is initialized.
 	 */
 	onInit() {
```
Comment (AlexDM0) [inline drivers/crownstone/driver.js:16]: Why is this file not .ts?

File: drivers/crownstone/driver.js [hunk 4]
```diff
@@ -121,37 +112,36 @@ class CrownstoneDriver extends Homey.Driver {
 		session.setHandler('list_devices', async (data) => {
 			let deviceList = [];
```
Comment (AlexDM0) [inline drivers/crownstone/driver.js:113]: Why does this not use the cache?

**Review by AlexDM0 — COMMENTED** [review]
Looks good! Main comment would be that the composition of modules seems a little strange for me.

Cache 
I'd expect this to be your main database. This class should have an update method which will refresh its data with the cloud's data. Similarly, partial updates like updateCrownstones, updateLocations etc would partially update these entries from the cloud.

Syncer. (Rename to CloudUtil since it does not sync)
This can be a stateless utility object which wraps the cloud lib. It will return the cloud data, which the cache then then optionally map and put in the cache.

Mapper
Stateless utility object which translates a crownstoneCloudLocation to a homeyLocation etc.

[DeviceDriverInterface] This class can house all the methods on the app.ts from line 271 onwards.

File: app.ts [hunk 5]
```diff
@@ -0,0 +1,377 @@
...
+		// We will not retry at a later time.
+		if (!this.containsAccountSettings()) {
+			console.log('There are no account settings... Cannot log in');
+			return;
+		}
+
+		console.log('Login to servers');
+		await this.synchronizeCloud();
+
+		this.pollPresenceInterval = 30;
```
[the remark this answers is not in the document]
Comment (mrquincle) [reply app.ts:63]: I've not explicitly stated that this is two-way synchronization. It's just one-way. I will then call it `mirroring` to be more explicit on this. That's the term almost always used for one-way syncing.

File: source/mapper.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+		};
+		this.userList.unshift(anyUser);
+		console.log('List with users:', this.userList);
+		return this.userList;
+	}
+
+	/**
+	 * A helper function for extractUsers. This pushes all not-yet existing users to the given userList.
+	 */
+	addUsersToList(userList: homey_User[], users: cloud_UserData[]) {
```
[the remark this answers is not in the document]
Comment (mrquincle) [reply source/mapper.ts]: I'll then introduce a secondary cache that stores the entities that are expected by the Homey classes. That will make the function of the mapper obvious and stateless.

File: app.ts (hunk 2 above)
Comment (mrquincle) [reply app.ts]: Done, must be tested

File: drivers/crownstone/driver.js (hunk 4 above)
Comment (mrquincle) [reply drivers/crownstone/driver.js]: > Why does this not use the cache?

This needs some more love, but there are basically two routes to fill in the data structures in the app:

1. via app boot and 
2. via first time configuration after install (that is in `onPair`). 

In the `onPair()` route we should be able to list devices as fast as possible. This process takes actually quite long and keeps the user waiting. 

Other stuff w.r.t. users, presence etc. is fine to be loaded later. 

Of course, the Homey persists devices etc. over reboots. It's just trying to err on the side of safety to obtain stuff from our cloud on that moment.

**Pull request merged**